taichi/ir/frontend_ir.cpp

```diff
@@ -128,10 +131,12 @@ void UnaryOpExpression::serialize(std::ostream &ss) {
 }
 
 void UnaryOpExpression::type_check() {
-  // TODO: assert no unknowns after type_check for all expressions are
-  // implemented
-  if (operand->ret_type == PrimitiveType::unknown)
-    return;
+  TI_ASSERT_INFO(operand->ret_type != PrimitiveType::unknown,
```

Comment:
nit: maybe this assert worth its own macro :P

Reply:
Nice suggestion! Done.